File: master-backend/app/life/services/doctor_service.py
"""Doctor-portal service for LifeSevatra — Firestore backed.

Fixes applied:
- Replaced local ``_now_iso()`` with ``now_iso()`` from ``firebase_client``.
- ``get_doctor_patients``: queries ``doctor_id`` (the field written by
  ``admission_service``) instead of the non-existent ``assigned_doctor_id``.
"""

import logging
from typing import Optional

from app.firebase_client import now_iso

logger = logging.getLogger(__name__)


# ━━━━━━━━━━━━━━━━━━━━━━━━ Patients ━━━━━━━━━━━━━━━━━━━━━━━━


async def get_doctor_patients(
    db, staff_doc_id: str, hospital_id: str
) -> list[dict]:
    """Return all admissions where the assigned doctor matches this staff member.

    FIX: ``admission_service.create_admission`` writes the field as
    ``doctor_id``, so we query on that — not ``assigned_doctor_id``.
    """
    docs = (
        db.collection("life_admissions")
        .where("hospital_id", "==", hospital_id)
        .where("doctor_id", "==", staff_doc_id)
        .get()
    )
    results = []
    for d in docs:
        data = d.to_dict()
        data["id"] = d.id
        results.append(data)
    return results


# ━━━━━━━━━━━━━━━━━━━━━━━━ Schedule ━━━━━━━━━━━━━━━━━━━━━━━━


async def get_schedule(db, staff_doc_id: str) -> list[dict]:
    """Get schedule entries for a doctor."""
    docs = (
        db.collection("life_schedules")
        .where("doctor_id", "==", staff_doc_id)
        .order_by("time")
        .get()
    )
    return [{"id": d.id, **d.to_dict()} for d in docs]


async def create_schedule_slot(
    db, staff_doc_id: str, hospital_id: str, data: dict
) -> dict:
    now = now_iso()
    doc_data = {
        "doctor_id": staff_doc_id,
        "hospital_id": hospital_id,
        "time": data.get("time", ""),
        "patient_name": data.get("patient_name", ""),
        "patient_id": data.get("patient_id"),
        "type": data.get("type", "checkup"),
        "status": "scheduled",
        "notes": data.get("notes", ""),
        "created_at": now,
        "updated_at": now,
    }
    doc_ref = db.collection("life_schedules").document()
    doc_ref.set(doc_data)
    doc_data["id"] = doc_ref.id
    return doc_data


async def update_schedule_status(
    db, schedule_id: str, staff_doc_id: str, new_status: str
) -> dict:
    ref = db.collection("life_schedules").document(schedule_id)
    doc = ref.get()
    if not doc.exists:
        return None
    data = doc.to_dict()
    if data.get("doctor_id") != staff_doc_id:
        return None
    ref.update({"status": new_status, "updated_at": now_iso()})
    data["status"] = new_status
    data["id"] = doc.id
    return data


# ━━━━━━━━━━━━━━━━━━━━━━━━ Clinical Notes ━━━━━━━━━━━━━━━━━━━━━━━━


async def get_clinical_notes(
    db,
    staff_doc_id: str,
    patient_id: Optional[str] = None,
    note_type: Optional[str] = None,
) -> list[dict]:
    """Return clinical notes written by this doctor."""
    query = db.collection("life_clinical_notes").where(
        "doctor_id", "==", staff_doc_id
    )
    if patient_id:
        query = query.where("patient_id", "==", patient_id)
    if note_type:
        query = query.where("type", "==", note_type)
    docs = query.get()
    return [{"id": d.id, **d.to_dict()} for d in docs]


async def add_clinical_note(
    db, staff_doc_id: str, doctor_name: str, data: dict
) -> dict:
    now = now_iso()
    doc_data = {
        "doctor_id": staff_doc_id,
        "doctor_name": doctor_name,
        "patient_id": data["patient_id"],
        "patient_name": data["patient_name"],
        "note": data["note"],
        "type": data.get("type", "progress"),
        "created_at": now,
    }
    doc_ref = db.collection("life_clinical_notes").document()
    doc_ref.set(doc_data)
    doc_data["id"] = doc_ref.id
    return doc_data


# ━━━━━━━━━━━━━━━━━━━━━━━━ Profile ━━━━━━━━━━━━━━━━━━━━━━━━


async def get_doctor_profile(db, staff_doc_id: str) -> Optional[dict]:
    ref = db.collection("life_staff").document(staff_doc_id)
    doc = ref.get()
    if not doc.exists:
        return None
    data = doc.to_dict()
    data["id"] = doc.id
    return data


async def update_doctor_profile(
    db, staff_doc_id: str, updates: dict
) -> Optional[dict]:
    ref = db.collection("life_staff").document(staff_doc_id)
    doc = ref.get()
    if not doc.exists:
        return None

    allowed = {
        "full_name",
        "specialty",
        "qualification",
        "experience_years",
        "contact",
        "email",
        "bio",
        "languages",
        "consultation_fee",
        "shift",
    }
    clean = {k: v for k, v in updates.items() if k in allowed and v is not None}
    clean["updated_at"] = now_iso()

    ref.update(clean)
    updated = ref.get().to_dict()
    updated["id"] = staff_doc_id
    return updated

File: lifesavatra-frontend/src/services/doctorService.ts
/**
 * Doctor Service
 * Doctor-portal specific API calls: my patients, schedule, clinical notes, profile.
 * Calls the FastAPI backend at /api/doctor/*.
 */

import { getAuthHeaders } from '../context/AuthContext';
import type { AdmittedPatient } from './admissionService';
export type { AdmittedPatient } from './admissionService';

const API = import.meta.env.VITE_API_URL ?? 'https://api-sevatra.vercel.app/api/v1';

// ── Interfaces ──────────────────────────────────────────────────────────────

export interface ScheduleSlot {
  id: string;
  time: string;
  patient_name: string;
  patient_id?: string;
  type: string;
  status: 'scheduled' | 'completed' | 'in-progress' | 'cancelled' | 'no-show';
  notes: string;
}

export interface ClinicalNote {
  id: string;
  patient_id: string;
  patient_name: string;
  doctor_id?: string;
  doctor_name?: string;
  note: string;
  created_at: string;
  type: 'observation' | 'prescription' | 'discharge-summary' | 'progress';
}

export interface DoctorProfile {
  id: string;
  staff_id: string;
  full_name: string;
  email?: string;
  role: string;
  specialty: string;
  qualification?: string | null;
  experience_years: number;
  contact?: string;
  on_duty: boolean;
  shift: string;
  max_patients: number;
  current_patient_count: number;
  hospital_id: string;
  joined_date: string;
  bio?: string;
  languages?: string[];
  consultation_fee?: number;
}

// ── Helpers ──────────────────────────────────────────────────────────────────

async function api<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${API}${path}`, {
    headers: { 'Content-Type': 'application/json', ...getAuthHeaders(), ...init?.headers },
    ...init,
  });
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    throw new Error(body.detail ?? `API error ${res.status}`);
  }
  return res.json();
}

// ── Patients ─────────────────────────────────────────────────────────────────

export const getDoctorPatients = async (): Promise<AdmittedPatient[]> => {
  const res = await api<{ patients: AdmittedPatient[]; total: number }>('/life/doctor/patients');
  return res.patients;
};

export const getPatientById = async (patientId: string): Promise<AdmittedPatient> =>
  api<AdmittedPatient>(`/life/doctor/patients/${patientId}`);

export const updatePatientVitals = async (
  patientId: string,
  vitals: { heartRate?: number; spo2?: number; respRate?: number; temperature?: number; bpSystolic?: number; bpDiastolic?: number },
): Promise<AdmittedPatient> => api<AdmittedPatient>(`/life/admissions/${patientId}/vitals`, { method: 'PUT', body: JSON.stringify(vitals) });

export const updatePatientClinicalInfo = async (
  patientId: string,
  info: { clinicalNotes?: string; labResults?: string },
): Promise<AdmittedPatient> => api<AdmittedPatient>(`/life/admissions/${patientId}/clinical`, { method: 'PUT', body: JSON.stringify(info) });

// ── Schedule ─────────────────────────────────────────────────────────────────

export const getSchedule = async (): Promise<ScheduleSlot[]> => {
  const res = await api<{ schedule: ScheduleSlot[] }>('/life/doctor/schedule');
  return res.schedule;
};

export const createScheduleSlot = async (data: {
  time: string;
  patient_name?: string;
  patient_id?: string;
  type: string;
  notes?: string;
  status?: string;
}): Promise<ScheduleSlot> =>
  api<ScheduleSlot>('/life/doctor/schedule', { method: 'POST', body: JSON.stringify(data) });

export const updateScheduleStatus = async (slotId: string, status: string): Promise<ScheduleSlot> =>
  api<ScheduleSlot>(`/life/doctor/schedule/${slotId}`, { method: 'PUT', body: JSON.stringify({ status }) });

// ── Clinical Notes ───────────────────────────────────────────────────────────

export const getClinicalNotes = async (patientId?: string, type?: string): Promise<ClinicalNote[]> => {
  const params = new URLSearchParams();
  if (patientId) params.set('patient_id', patientId);
  if (type) params.set('type', type);
  const qs = params.toString() ? `?${params}` : '';
  const res = await api<{ notes: ClinicalNote[]; total: number }>(`/life/doctor/notes${qs}`);
  return res.notes;
};

export const addClinicalNote = async (note: {
  patient_id: string;
  patient_name: string;
  note: string;
  type: string;
}): Promise<ClinicalNote> =>
  api<ClinicalNote>('/life/doctor/notes', { method: 'POST', body: JSON.stringify(note) });

// ── Profile ──────────────────────────────────────────────────────────────────

export const getDoctorProfile = async (): Promise<DoctorProfile> =>
  api<DoctorProfile>('/life/doctor/profile');

export const updateDoctorProfile = async (updates: Partial<DoctorProfile>): Promise<DoctorProfile> =>
  api<DoctorProfile>('/life/doctor/profile', { method: 'PUT', body: JSON.stringify(updates) });
